File: app/src/main/java/chat/rocket/android/helper/LoadMoreScrollListener.java
package chat.rocket.android.helper;

import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;

@SuppressWarnings("PMD.AbstractNaming")
public abstract class LoadMoreScrollListener extends RecyclerView.OnScrollListener {

  private final LinearLayoutManager layoutManager;
  private final int loadThreshold;
  private boolean isLoading;

  /**
   * constructor. loadThreshold is better to set to 0.4 * total.
   */
  public LoadMoreScrollListener(LinearLayoutManager layoutManager, int loadThreshold) {
    this.layoutManager = layoutManager;
    this.loadThreshold = loadThreshold;
    setLoadingDone();
  }

  @Override
  public void onScrolled(RecyclerView recyclerView, int deltaX, int deltaY) {
    super.onScrolled(recyclerView, deltaX, deltaY);

    final int visibleItemCount = recyclerView.getChildCount();
    final int totalItemCount = layoutManager.getItemCount();
    final int firstVisibleItem = layoutManager.findFirstVisibleItemPosition();

    if (!isLoading
        && firstVisibleItem + visibleItemCount >= totalItemCount - loadThreshold
        && visibleItemCount < totalItemCount
        && deltaY < 0) {
      isLoading = true;
      requestMoreItem();
    }
  }

  public void setLoadingDone() {
    isLoading = false;
  }

  public abstract void requestMoreItem();
}

File: app/src/main/java/chat/rocket/android/helper/LoadMoreScrollListener.kt
package chat.rocket.android.helper

import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView

abstract class LoadMoreScrollListener(private val layoutManager: LinearLayoutManager, private val visibleThreshold: Int) : RecyclerView.OnScrollListener() {
    private var isLoading: Boolean = false

    override fun onScrolled(recyclerView: RecyclerView, deltaX: Int, deltaY: Int) {
        super.onScrolled(recyclerView, deltaX, deltaY)

        val visibleItemCount = recyclerView.childCount
        val totalItemCount = layoutManager.itemCount
        val firstVisibleItem = layoutManager.findFirstVisibleItemPosition()

        if (!isLoading
                && (firstVisibleItem + visibleItemCount) >= (totalItemCount - visibleThreshold)
                && visibleItemCount < totalItemCount
                && deltaY < 0) {
            isLoading = true
            requestMoreItem()
        }
    }

    fun setLoadingDone() {
        isLoading = false
    }

    abstract fun requestMoreItem()
}